core/modules/css/KvInterfaceImplMySql.cc

```diff
@@ -0,0 +1,319 @@
+#include "sql/SqlResults.h"
+#include "sql/SqlTransaction.h"
+
+using std::map;
+using std::string;
+using std::vector;
+
+namespace {
+
+// logger instance for this module
+LOG_LOGGER _logger = LOG_GET("lsst.qserv.css.KvInterfaceImplMySql");
+
+}
+
+
+namespace lsst {
+namespace qserv {
+namespace css {
+
+// todo this is copied more or less directly from QMetaTransaction. Refactor to shared location?
+// proper place is to port it into db module
+// https://github.com/LSST/db
+// right now it's python only.
+// create a ticket to make generalized code to be used between c++ and python
+class KvTransaction
+{
+public:
+    // Constructors
+    KvTransaction(sql::SqlConnection& conn)
+        : _errObj(), _trans(conn, _errObj) {
+        if (_errObj.isSet()) {
+            throw CssError(_errObj);
+        }
+    }
+
+    // Destructor
+    ~KvTransaction() {
+        // instead of just destroying SqlTransaction instance we call abort and see
+        // if error happens. We cannot throw here but we can print a message.
+        if (_trans.isActive()) {
+            _trans.abort(_errObj);
+            if (_errObj.isSet()) {
+                LOGF(_logger, LOG_LVL_ERROR, "Failed to abort transaction: mysql error: (%1%) %2%" %
+                     _errObj.errNo() % _errObj.errMsg());
+            }
+        }
+    }
+
+    /// Explicitly commit transaction, throws SqlError for errors.
+    void commit() {
+        _trans.commit(_errObj);
+        if (_errObj.isSet()) {
+            LOGF(_logger, LOG_LVL_ERROR, "Failed to commit transaction: mysql error: (%1%) %2%" %
+                 _errObj.errNo() % _errObj.errMsg());
+            throw CssError(_errObj);
+        }
+    }
+
+    /// Explicitly abort transaction, throws SqlError for errors.
+    void abort() {
+        _trans.abort(_errObj);
+        if (_errObj.isSet()) {
+            LOGF(_logger, LOG_LVL_ERROR, "Failed to abort transaction: mysql error: (%1%) %2%" %
+                 _errObj.errNo() % _errObj.errMsg());
+            throw CssError(_errObj);
+        }
+    }
+
+private:
+    sql::SqlErrorObject _errObj; // this must be declared before _trans
+    sql::SqlTransaction _trans;
+};
+
+
+KvInterfaceImplMySql::KvInterfaceImplMySql(mysql::MySqlConfig const& mysqlConf)
+: _conn(mysqlConf) {
+
+}
+
+
+void
+KvInterfaceImplMySql::create(std::string const& key, std::string const& value) {
+    if (key.size() > kMaxKeyLength) {
+        LOGF(_logger, LOG_LVL_DEBUG, "create - rejecting; key too long: %1%" % key);
+        throw CssError("key too long");
+    }
+
+    KvTransaction transaction(_conn);
+
+    boost::filesystem::path keyPathObj(key);
+    boost::filesystem::path parentKeyPathObj = keyPathObj.parent_path();
+
+    std::string parentKey = boost::filesystem::path(key).parent_path().string();
+
+    boost::format fmQuery("INSERT INTO cssData (kvKey, kvVal, parentKey) VALUES ('%1%', '%2%', '%3%');");
+    fmQuery % key;
+    fmQuery % value;
+    fmQuery % parentKey;
```

Comment:
What about Bobby Tables key? :)

You need to properly escape all string values with `_conn.escapeString(string)`. And I would not use boost format here, it's kind of expensive, just concatenating strings is more light-weight.